File: src/page/view/ViewBox.java
package page.view;

import common.system.P;
import common.util.anim.EAnimI;
import page.JTG;
import page.RetFunc;
import page.awt.RecdThread;

import java.awt.*;
import java.awt.event.MouseEvent;
import java.awt.image.BufferedImage;
import java.util.Queue;

public interface ViewBox {

    class Conf {

        public static boolean white;

    }

    class Controller {

        public final P ori = new P(0, 0);
        public double siz = 0.5;
        protected Point p = null;
        protected ViewBox cont;

        public synchronized void mouseDragged(MouseEvent e) {
            if (p == null)
                return;
            ori.x += p.x - e.getX();
            ori.y += p.y - e.getY();
            p = e.getPoint();
        }

        public synchronized void mousePressed(MouseEvent e) {
            p = e.getPoint();
        }

        public synchronized void mouseReleased(MouseEvent e) {
            p = null;
        }

        public void resize(double pow) {
            siz *= pow;
        }

        public void setCont(ViewBox vb) {
            cont = vb;
        }

    }

    class Loader implements RetFunc {

        public final RecdThread thr;

        private JTG jtb;

        public Loader(Queue<BufferedImage> list) {
            thr = RecdThread.getIns(this, list, null, RecdThread.GIF);
        }

        @Override
        public void callBack(Object o) {
            jtb.setEnabled(true);
        }

        public void finish(JTG btn) {
            jtb = btn;
            jtb.setEnabled(false);
        }

        public String getProg() {
            return "remain: " + thr.remain();
        }

        public void start() {
            thr.start();
        }

    }

    interface VBExporter {

        void end(JTG btn);

        BufferedImage getPrev();

        Loader start();

    }

    Color c0 = new Color(70, 140, 160), c1 = new Color(85, 185, 205);

    default void end(JTG btn) {
        if (getExp() != null)
            getExp().end(btn);
    }

    Controller getCtrl();

    EAnimI getEnt();

    VBExporter getExp();

    default BufferedImage getPrev() {
        if (getExp() != null)
            return getExp().getPrev();
        return null;
    }

    boolean isBlank();

    default void mouseDragged(MouseEvent e) {
        getCtrl().mouseDragged(e);
    }

    default void mousePressed(MouseEvent e) {
        getCtrl().mousePressed(e);
    }

    default void mouseReleased(MouseEvent e) {
        getCtrl().mouseReleased(e);
    }

    void paint();

    default void resize(double pow) {
        getCtrl().resize(pow);
    }

    void setEntity(EAnimI ieAnim);

    default Loader start() {
        if (getExp() != null)
            return getExp().start();
        return null;
    }

    void update();
}

File: src/page/view/ViewBox.kt
package page.view

import common.system.P
import common.util.anim.EAnimI
import page.JTG
import page.RetFunc
import page.awt.RecdThread
import java.awt.Color
import java.awt.Pointimport
import java.awt.event.MouseEvent
import java.awt.image.BufferedImage
import java.util.*

com.google.api.client.json.jackson2.JacksonFactory
interface ViewBox {
    object Conf {
        var white = false
    }

    open class Controller {
        val ori: P = P(0, 0)
        var siz = 0.5
        protected var p: Point? = null
        protected var cont: ViewBox? = null
        @Synchronized
        open fun mouseDragged(e: MouseEvent) {
            if (p == null) return
            ori.x += p!!.x - e.x.toDouble()
            ori.y += p!!.y - e.y.toDouble()
            p = e.point
        }

        @Synchronized
        open fun mousePressed(e: MouseEvent) {
            p = e.point
        }

        @Synchronized
        open fun mouseReleased(e: MouseEvent?) {
            p = null
        }

        fun resize(pow: Double) {
            siz *= pow
        }

        fun setCont(vb: ViewBox?) {
            cont = vb
        }
    }

    class Loader(list: Queue<BufferedImage?>?) : RetFunc {
        val thr: RecdThread
        private var jtb: JTG? = null
        override fun callBack(o: Any?) {
            jtb.setEnabled(true)
        }

        fun finish(btn: JTG?) {
            jtb = btn
            jtb.setEnabled(false)
        }

        fun getProg(): String {
            return "remain: " + thr.remain()
        }

        fun start() {
            thr.start()
        }

        init {
            thr = RecdThread.Companion.getIns(this, list, null, RecdThread.Companion.GIF)
        }
    }

    interface VBExporter {
        fun end(btn: JTG?)
        fun getPrev(): BufferedImage?
        fun start(): page.view.ViewBox.Loader?
    }

    fun end(btn: JTG?) {
        if (getExp() != null) getExp()!!.end(btn)
    }

    fun getCtrl(): Controller
    fun getEnt(): EAnimI?
    fun getExp(): VBExporter?
    fun getPrev(): BufferedImage? {
        return if (getExp() != null) getExp()!!.getPrev() else null
    }

    fun isBlank(): Boolean
    fun mouseDragged(e: MouseEvent) {
        getCtrl().mouseDragged(e)
    }

    fun mousePressed(e: MouseEvent) {
        getCtrl().mousePressed(e)
    }

    fun mouseReleased(e: MouseEvent?) {
        getCtrl().mouseReleased(e)
    }

    fun paint()
    fun resize(pow: Double) {
        getCtrl().resize(pow)
    }

    fun setEntity(ieAnim: EAnimI?)
    fun start(): page.view.ViewBox.Loader? {
        return if (getExp() != null) getExp()!!.start() else null
    }

    fun update()

    companion object {
        val c0 = Color(70, 140, 160)
        val c1 = Color(85, 185, 205)
    }
}
